File: src/main/scala/com/dylowen/tinglebot/brain/TextBrain.java
package com.dylowen.tinglebot.brain;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

/**
 * TODO add description
 *
 * @author dylan.owen
 * @since Feb-2016
 */
public class TextBrain
    extends Brain<String, String> {

    private transient static final int MAX_WORDS_IN_SENTENCE = 100;

    public TextBrain(final int gramSize) {
        super(gramSize);
    }

    @Override
    protected String getNextWord(final LinkedList<String> input) {
        final String nextWord = super.getNextWord(input);

        return (nextWord == null) ? "." : nextWord;
    }

    @Override
    public String concatSentence(final List<String> words) {
        final StringBuilder sb = new StringBuilder();
        final Iterator<String> it = words.listIterator();

        GenericWordType.START_SENTENCE.appendSpacedWord(it.next(), sb);

        while (it.hasNext()) {
            GenericWordType.appendUntypedWord(it.next(), sb);
        }

        return sb.toString();
    }

    @Override
    protected boolean shouldContinueSentence(final String nextWord, final LinkedList<String> sentence) {
        return GenericWordType.getType(nextWord) != GenericWordType.END_SENTENCE
                && sentence.size() < MAX_WORDS_IN_SENTENCE;
    }
}

File: src/main/scala/com/dylowen/tinglebot/brain/TextBrain.scala
package com.dylowen.tinglebot.brain

/**
  * TODO add description
  *
  * @author dylan.owen
  * @since Mar-2016
  */
abstract class TextBrain(gramSize: Integer) extends Brain[String, String](gramSize) {

  @transient
  private val MAX_WORDS_IN_SENTENCE: Integer = 50

  //if we can't continue, inject a period
  override def getNextWord(input: List[String]): Option[String] = Option(super.getNextWord(input).getOrElse("."))

  override def concatSentence(words: List[String]): String = words.foldLeft(new StringBuilder())((builder, word) =>
    builder.append(GenericWordType.getSpacedWord(word))
  ).toString()

  override def shouldContinueSentence(word: String, sentence: List[String]): Boolean = {
    GenericWordType.getType(word) != GenericWordType.END_SENTENCE && sentence.size < MAX_WORDS_IN_SENTENCE
  }
}
